lib/authentication.js

```diff
@@ -0,0 +1,172 @@
+'use strict';
+
+const MAX_PASS_LENGTH = 128;
+const MIN_PASS_LENGTH = 10;
+const MIN_PHRASE_LENGTH = 20;
+
+const MIN_LOGIN_LENGTH = 6;
+const MAX_LOGIN_LENGTH = 255;
+
+const passwordTests = {
+  MIN_LENGTH: [
+    password => password.length < MIN_PASS_LENGTH,
+    `The password must be longer than ${MIN_PASS_LENGTH} characters`
+  ],
+  MAX_LENGTH: [
+    password => password.length > MAX_PASS_LENGTH,
+    `The password must be less than ${MAX_PASS_LENGTH} characters`
+  ],
+  MIN_PHRASE_LENGTH: [
+    password => password.length < MIN_PHRASE_LENGTH,
+    `The passphrase must be longer than ${MIN_PHRASE_LENGTH} characters`
+  ],
+  REPEAT_CHARS: [
+    password => /(.)\1{2,}/.test(password),
+    'The password can not contain a sequence of repeated chars'
+  ],
+  ONE_LOWERCASE_CHAR: [
+    password => !/[a-z]/.test(password),
+    'The password must contain lowercase char'
+  ],
+  ONE_UPPERERCASE_CHAR: [
+    password => !/[A-Z]/.test(password),
+    'The password must contain uppercase char'
+  ],
+  ONE_NUMBER: [
+    password => !/[0-9]/.test(password),
+    'The password must contain number'
+  ],
+  ONE_SPECIAL_CHAR: [
+    password => !/[^A-Za-z0-9]/.test(password),
+    'The password must contain special char'
+  ]
+};
+
+const loginTests = {
+  MIN_LENGTH: [
+    login => login.length < MIN_LOGIN_LENGTH,
+    `The login must be longer than ${MIN_LOGIN_LENGTH} characters`
+  ],
+  MAX_LENGTH: [
+    login => login.length > MAX_LOGIN_LENGTH,
+    `The login must be less than ${MAX_LOGIN_LENGTH} characters`
+  ]
+};
+
+const loginPasswordTests = {
+  LOGIN_INCLUDES_PASSWORD: [
+    (login, password) => login.includes(password),
+    'The login can not contain a password'
+  ],
+  PASSWORD_INCLUDES_LOGIN: [
+    (login, password) => password.includes(login),
+    'The password can not contain a login'
+  ]
+};
+
+class AuthenticationStrength {
+  constructor(
+    hints, // Array, [{ required: Array, optional: Array }, ...]
+    bestPercent // Number, percent of best test group
+  ) {
+    this.hints = hints;
+    this.bestPercent = bestPercent;
+  }
+  get valid() {
+    return !!this.bestPercent;
+  }
+  get strength() {
+    if (this.bestPercent < 25) {
+      return 'Very weak';
+    } else if (this.bestPercent < 40) {
+      return 'Weak';
+    } else if (this.bestPercent < 60) {
+      return 'Good';
+    } else if (this.bestPercent < 80) {
+      return 'Strong';
+    } else {
+      return 'Very strong';
+    }
+  }
+}
+
+const makeTest = (
+  // Function that checks the arguments on a test suite
+  tests, // Array, array of password/login tests
+  testGroups, // Array, [{ required: [testNames], optional: [testNames] }]
+  ...testArgs // Array, [password]/[login]/[login, password]
```

Comment:
Could you please reorder it to
`[login] / [password] / [login, password]`
?
=)